行数がまともじゃない

diffhunk を見てレビューコメントしたい位置までの -行も含めた行数でpositionを決定しないとまともな位置がでないっぽい！

これもなんか使えそうかな hunk
```
❯ git diff -U0 master commands/review-comments.sh
diff --git a/commands/review-comments.sh b/commands/review-comments.sh
index f946345..25e9779 100755
--- a/commands/review-comments.sh
+++ b/commands/review-comments.sh
@@ -5,9 +5,2 @@ review-comments() {
-  echo "[{\"path\": \"$1\", \"position\": $2, \"body\": \"$3\"}]" > TMP.json
-  REVIEW_COMMENT_STATE="./.REVIEW_COMMENT_STATE"
-  if [ -e $REVIEW_COMMENT_STATE ]; then
-    # shellcheck disable=SC2094
-    jq -s add "$REVIEW_COMMENT_STATE" TMP.json "." > NEW_REVIEW_COMMENT_STATE
-    mv NEW_REVIEW_COMMENT_STATE "$REVIEW_COMMENT_STATE"
-  else
-    cp TMP.json "$REVIEW_COMMENT_STATE"
-  fi
+  local REVIEW_COMMENT_STATE="./.REVIEW_COMMENT_STATE"
+  local current=$(cat -- "$REVIEW_COMMENT_STATE" 2> /dev/null)
@@ -15 +8,6 @@ review-comments() {
-  rm TMP.json
+  jq \
+    --arg path "$1" \
+    --arg position "$2" \
+    --arg body "$3" \
+    'flatten + [{path: $path, position: $position | tonumber, body: $body}]' \
+    <<< "${current:-[]}" > "$REVIEW_COMMENT_STATE"
```

--- a/commands/line-to-position.sh
+++ b/commands/line-to-position.sh
@@ -90,7 +90,7 @@ line-in-range() {
       sum = $0
       gsub(/.*,/, "", sum)
 
-      for (i = start; i <= start + sum; ++i) {
+      for (i = int(start); i <= int(start) + int(sum); ++i) {
         if (i == line) {
           found = 1
           exit

--- a/commands/line-to-position.sh
+++ b/commands/line-to-position.sh
@@ -23,10 +23,12 @@ line-to-position() {
   awk \
     -v line="$line" \
     -v target="$target" '
-    NR <= 4 {
+    !hunk && !/^@@/ {
       next
     }
     /^@@/ {
+      hunk = 1
+
       if (target == "add") {
         gsub(/^@@[^+]+\+| @@.*/, "")
       }

--- a/commands/line-to-position.sh
+++ b/commands/line-to-position.sh
@@ -1,0 +1,108 @@
+#!/usr/bin/env bash
+
+# line-to-position 1 add
+line-to-position() {
+  local diff=$(cat)
+  local line=$1
+  local target=$2
+  shift 2
+
+  if [[ ! $line =~ ^[0-9]+$ ]]; then
+    echo '$1 must be a number' >&2
+    return 1
+  fi
+  if [[ ! $target =~ ^(add|remove)$ ]]; then
+    echo '$2 must be add or remove' >&2
+    return 1
+  fi
+
+  if ! line-in-range "$line" "$target" <<< "$diff"; then
+    return 1
+  fi
+
+  awk \
+    -v line="$line" \
+    -v target="$target" '
+    NR <= 4 {
+      next
+    }
+    /^@@/ {
+      if (target == "add") {
+        gsub(/^@@[^+]+\+| @@.*/, "")
+      }
+      if (target == "remove") {
+        gsub(/^@@ -| .*/, "")
+      }
+
+      if (current > 1) {
+        ++position
+      }
+      start = $0
+      gsub(/,.*/, "", start)
+      current = start - 1
+    }
+    /^ / {
+      ++position
+      ++current
+    }
+    /^\+/ {
+      ++position
+      if (target == "add") {
+        ++current
+      }
+    }
+    /^-/ {
+      ++position
+      if (target == "remove") {
+        ++current
+      }
+    }
+    current == line {
+      print position
+      exit
+    }
+  ' <<< "$diff"
+}
+
+# line-in-range 1 add
+line-in-range() {
+  local line=$1
+  local target=$2
+  shift 2
+
+  awk \
+    -v line="$line" \
+    -v target="$target" '
+    /^@@/ {
+      orig = $0
+      gsub(/ @@.*/, " @@", orig)
+
+      if (target == "add") {
+        gsub(/^@@[^+]+\+| @@.*/, "")
+      }
+      if (target == "remove") {
+        gsub(/^@@ -| .*/, "")
+      }
+
+      start = $0
+      gsub(/,.*/, "", start)
+
+      sum = $0
+      gsub(/.*,/, "", sum)
+
+      for (i = start; i <= start + sum; ++i) {
+        if (i == line) {
+          found = 1
+          exit
+        }
+      }
+    }
+    END {
+      if (!found) {
+        exit 1
+      }
+    }
+  '
+}
+
+line-to-position "$@"

--- a/commands/review-comments.sh
+++ b/commands/review-comments.sh
@@ -11,7 +11,7 @@ review-comments() {
 
   # TODO: Change upstream branch (currently `master`) to correct one!
   # TODO: Change `add` to correct one!
-  if ! position=$(git diff --diff-algorithm=default master -- "$path" | "$(dirname -- "$0")"/line-to-position "$line" add); then
+  if ! position=$(git diff --diff-algorithm=default master -- "$path" | "$(dirname -- "$0")"/line-to-position.sh "$line" add); then
     return 1
   fi
 

--- a/commands/review-comments.sh
+++ b/commands/review-comments.sh
@@ -4,11 +4,21 @@
 review-comments() {
   local REVIEW_COMMENT_STATE="./.REVIEW_COMMENT_STATE"
   local current=$(cat -- "$REVIEW_COMMENT_STATE" 2> /dev/null)
+  local path=$1
+  local line=$2
+  local body=$3
+  local position
+
+  # TODO: Change upstream branch (currently `master`) to correct one!
+  # TODO: Change `add` to correct one!
+  if ! position=$(git diff --diff-algorithm=default master -- "$path" | "$(dirname -- "$0")"/line-to-position "$line" add); then
+    return 1
+  fi
 
   jq \
-    --arg path "$1" \
-    --arg position "$2" \
-    --arg body "$3" \
+    --arg path "$path" \
+    --arg position "$position" \
+    --arg body "$body" \
     'flatten + [{path: $path, position: $position | tonumber, body: $body}]' \
     <<< "${current:-[]}" > "$REVIEW_COMMENT_STATE"
 }

